### ISSUE TITLE
TCP, ENFILE, and log spam

### ISSUE THREAD

**Issue by blblack** [gdnsd/gdnsd#197]
The gdnsd instances at the WMF experienced an annoying problem related to fd ulimits - 

The TL;DR is that our configuration allowed for more TCP client threads than the daemon's file descriptor ulimit would support, and when we finally got enough connections to hit the ulimit, this caused an insane amount of syslog spam which filled a local disk ( https://phabricator.wikimedia.org/T266746 ).

There are a few different ways gdnsd could've handled the situation better than it did, and any one of them might've kept this problem from being as dramatic as it was:

1. The too-small ulimit was merely a soft ulimit in this case, while the hard limit was sufficiently-high to have avoided the problem.  gdnsd could've at least attempted to raise the soft limit by default, based on the configured `tcp_threads` and `tcp_clients_per_thread` values and such.
2. The `EMFILE` and `ENFILE` errnos for `accept4` (which is what we get when the ulimit is too small) can be very spammy if they're not handled in some way.  If you just log about them and move back to the eventloop, the pending connection persists and `accept4` will keep looping back to the same errno for the same connection.  We should handle this like we do our existing software-level connection limits and instead shoot down the most-idle connection, if possible, to make room.
3. Network-driven log outputs should probably have a basic-sanity level of ratelimiting in their output in the general case.
4. There's no stats output which tracks `accept4()` failures, meaning they'll never show up in monitoring or graphs, only in log outputs.  If the log output hadn't been spammy enough to fill a disk, we might not have otherwise noticed the issue at all!
